### PR TITLE
push for review

### PR DESCRIPTION
**Pull request by Ggoering**

### CODE REVIEW

File: src/main/java/com/GG/T9AgeCombat/predicates/CheckLimitationPredicate.java [hunk 1]
```diff
@@ -12,6 +12,6 @@ public static boolean checkFirstRound(Boolean isFirstRound) {
     }
 
     public static boolean checkHorde(Integer unitWidth) {
-        return unitWidth >= 10;
+        return unitWidth >= 8;
```
Comment (Arcontos) [inline src/main/java/com/GG/T9AgeCombat/predicates/CheckLimitationPredicate.java]: I think I added a value to the Constants file for hordes.

File: src/main/java/com/GG/T9AgeCombat/service/repository/UnitService.java [hunk 2]
```diff
@@ -15,7 +26,132 @@ public UnitService(EntityManager entityManager) {
...
+                "FROM special_rule sr \n"+
+                "INNER JOIN unit_special_rule usr \n"+
+                "ON usr.special_rule_id = sr.id \n"+
+                "INNER JOIN modification m ON m.id = sr.modification_id \n"+
+                "INNER JOIN limitation l ON l.id = sr.limitation_id \n"+
+                "INNER JOIN timing t ON t.id = sr.timing_id \n"+
+                "WHERE usr.unit_id = " + unitId + "  \n"
+                , SpecialRuleEntity.class);
+
+        List specialRulesList = q2.getResultList();
```
Comment (Arcontos) [inline src/main/java/com/GG/T9AgeCombat/service/repository/UnitService.java]: I think each of these should be a separate method and then we can have a wrapper method that calls each of the separate queries and builds the Unit in its entirety.

File: src/main/resources/db/migration/V20200429153203429__baseline.sql [hunk 3]
```diff
@@ -325,60 +325,74 @@ CREATE INDEX "fki_fk_character_mount_special_rule$special_rule" ON public.charac
```
Comment (Arcontos) [inline src/main/resources/db/migration/V20200429153203429__baseline.sql:325]: Minor but if we are using flyway to track migrations shouldn't these changes be in a new migration script rather than modifying the baseline?

File: src/main/java/com/GG/T9AgeCombat/models/Equipment.java [hunk 4]
```diff
@@ -1,28 +1,25 @@
...
 
 import javax.persistence.*;
 import java.util.Collection;
+import java.util.List;
 
 @Value
 @Builder
-@Entity
 @NoArgsConstructor(force = true)
 @AllArgsConstructor
```
Comment (Arcontos) [inline src/main/java/com/GG/T9AgeCombat/models/Equipment.java]: I don't think we need these on non-entities since we are not building them from a query but I could be wrong.

File: src/main/java/com/GG/T9AgeCombat/models/SpecialRule.java [hunk 5]
```diff
@@ -1,30 +1,45 @@
...
+
+    public SpecialRule specialRuleFromDTO(EquipmentDTO entity) {
+        return SpecialRule.builder()
+                .name(SpecialRuleEnum.valueOf(entity.getSpecialRuleName()))
+                .limitation(LimitationEnum.valueOf(entity.getLimitationValue()))
+                .modification(ModificationEnum.valueOf(entity.getModificationValue()))
+                .timing(TimingEnum.valueOf(entity.getTimingValue()))
+                .value(entity.getSpecialRuleValue())
+                .build();
+    }
```
Comment (Arcontos) [inline src/main/java/com/GG/T9AgeCombat/models/SpecialRule.java]: These seem like they should be in a SpecialRuleService class to me. Constructing brand new SpecialRules doesn't seem related to a specific SpecialRule object.

File: src/main/java/com/GG/T9AgeCombat/models/Unit.java [hunk 6]
```diff
@@ -169,6 +142,16 @@ private void adjustModelCountFromPendingWounds() {
...
+    public List<SpecialRule> specialRuleList(List<SpecialRuleEntity> specialRuleEntities) {
+            return specialRuleEntities.stream().map(x -> SpecialRule.builder()
+                    .name(SpecialRuleEnum.valueOf(x.getName()))
+                    .limitation(LimitationEnum.valueOf(x.getLimitationValue()))
+                    .modification(ModificationEnum.valueOf(x.getModificationValue()))
+                    .timing(TimingEnum.valueOf(x.getTimingValue()))
+                    .value(x.getValue())
+                    .build()).collect(toList());
+        }
+
```
Comment (Arcontos) [inline src/main/java/com/GG/T9AgeCombat/models/Unit.java]: Should this happen when we build the object so we don't have to build the list each time we request it? If not we should rename this method to `getSpecialRuleList`

File: src/main/resources/application.properties [hunk 7]
```diff
@@ -3,6 +3,6 @@ spring.datasource.hikari.connectionTimeout=20000
 spring.datasource.hikari.maximumPoolSize=5
 
 ## PostgreSQL
-spring.datasource.url=jdbc:postgresql://localhost:5432/T9AgeCombat
+spring.datasource.url=jdbc:postgresql://localhost:5432/postgres
```
Comment (Arcontos) [inline src/main/resources/application.properties:6]: This is fine by me, I had named my schema T9AgeCombat rather than the default but I have no problems with this.

File: src/main/java/com/GG/T9AgeCombat/entities/EquipmentDTO.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    @Id
+    long id;
+    String name;
+    String equipmentType;
+    @Id
+    String specialRuleName;
+    int specialRuleValue;
+    String modificationValue;
+    String limitationValue;
+    String timingValue;
```
Comment (Arcontos) [inline src/main/java/com/GG/T9AgeCombat/entities/EquipmentDTO.java:31]: With this setup is a unit going to have multiple instances of an equipment for each special rule on that equipment? Would we rather each equipment have a list of SpecialRules like the Unit does?

**Review by Arcontos — COMMENTED** [review]
If my comments don't make sense it's probably because I misread how the new stuff works and I just need it explained to me!

File: src/main/java/com/GG/T9AgeCombat/service/repository/UnitService.java (hunk 2 above)
Comment (Ggoering) [reply src/main/java/com/GG/T9AgeCombat/service/repository/UnitService.java]: I'm into that

File: src/main/java/com/GG/T9AgeCombat/models/Equipment.java (hunk 4 above)
Comment (Ggoering) [reply src/main/java/com/GG/T9AgeCombat/models/Equipment.java]: I will experiment

File: src/main/resources/db/migration/V20200429153203429__baseline.sql (hunk 3 above)
Comment (Ggoering) [reply src/main/resources/db/migration/V20200429153203429__baseline.sql:325]: Ok I wasn't sure if we came to a conclusion on how we wanted to use it.  Next time.

File: src/main/java/com/GG/T9AgeCombat/models/SpecialRule.java (hunk 5 above)
Comment (Ggoering) [reply src/main/java/com/GG/T9AgeCombat/models/SpecialRule.java]: That makes sense. Then inject it into unit service

File: src/main/java/com/GG/T9AgeCombat/models/Unit.java (hunk 6 above)
Comment (Ggoering) [reply src/main/java/com/GG/T9AgeCombat/models/Unit.java]: This was the attempt at trying to do the @builder override.  It didn't work since it actually shows the method as being unused.  I just forgot to delete it.  This method goes in SpecialRulesService too.

File: src/main/java/com/GG/T9AgeCombat/entities/EquipmentDTO.java (hunk 8 above)
Comment (Ggoering) [reply src/main/java/com/GG/T9AgeCombat/entities/EquipmentDTO.java:31]: No they won't. See Equipment.java for the model that actually goes on the Unit.   I'll move the builder method to an equipment service class though.

File: src/main/java/com/GG/T9AgeCombat/predicates/CheckLimitationPredicate.java (hunk 1 above)
Comment (Ggoering) [reply src/main/java/com/GG/T9AgeCombat/predicates/CheckLimitationPredicate.java]: oh nice didn't even think to look there xD

File: src/main/java/com/GG/T9AgeCombat/service/repository/UnitService.java [hunk 9]
```diff
@@ -1,21 +1,152 @@
 package com.GG.T9AgeCombat.service.repository;
```
Comment (Arcontos) [inline src/main/java/com/GG/T9AgeCombat/service/repository/UnitService.java:1]: We don't have to do this now necessarily but it probably makes sense to extend the `getSpecialRuleByUnitId` to also get the unit's special rules based on their height classification.

File: src/main/java/com/GG/T9AgeCombat/models/Unit.java (hunk 6 above)
Comment (Arcontos) [reply src/main/java/com/GG/T9AgeCombat/models/Unit.java]: Did you mean to leave this method in here still? I think if you were trying to override the Setter you need to follow this example https://www.baeldung.com/lombok-builder-custom-setter

**Review by Arcontos — APPROVED** [review]
